icubam/backoffice/handlers/users.py

```diff
@@ -41,109 +44,129 @@ def get(self):
-    return self.select_icus([user_icu.icu_id for user_icu in user.icus])
-
   @tornado.web.authenticated
   def get(self):
-    userid = self.get_query_argument('id', None)
-    user = None
-    if userid is not None:
-      user = self.db.get_user(userid)
+    user = self.db.get_user(self.get_query_argument('id', None))
+    user_icus = set([i.icu_id for i in user.icus]) if user is not None else []
+    user_micus = set(
+        [i.icu_id for i in user.managed_icus]) if user is not None else []
+    return self.do_render(user, user_icus, user_micus, error=False)
+
+  def do_render(self,
+                user: store.User,
+                icus: List[int],
+                managed_icus: List[int],
+                error=False):
+    """Render the form for a given user."""
 
     user = user if user is not None else store.User()
-    icus_dto = self.select_icus_for_user(user)
-    self.render("user.html", icus=icus_dto, user=user, error=False)
-
-  def error(self, user, icus):
-    self.render("user.html", icus=icus, user=user, error=True)
-
-  """Returns the ICUs we need to add to/remove from the user. """
-  def get_icus_to_store(self, user):
-    form_icus = self.get_body_arguments("icus", [])
-    form_icus = set([int(icu) for icu in form_icus])
-    user_icus = set([user_icu.icu_id for user_icu in user.icus])
-    add =  list(form_icus.difference(user_icus))
-    remove = list(user_icus.difference(form_icus))
-    return add, remove
-
-  """Selects the DTO comparing ALL icus against the ones from the form."""
-  def select_icus_for_error(self):
-    form_icus = self.get_body_arguments("icus", [])
-    return self.select_icus([int(icu) for icu in form_icus])
+    self.prepare_for_display(user)
+    options = sorted(self.get_options(), key=lambda icu: icu.name)
 
-  @tornado.web.authenticated
-  def post(self):
-    user_dict = {}
-    for key in ["name", "email", "telephone"]:
-      user_dict[key] = self.get_body_argument(key, None)
+    return self.render("user.html", options=options, user=user, icus=icus,
+                       managed_icus=managed_icus, error=error)
 
-    # Password is not required when editing the user.
-    password = self.get_body_argument("password", None)
+  def get_options(self):
+    if self.user.is_admin:
+      options = self.db.get_icus()
+    else:
+      options = self.user.managed_icus
+    return options
+
+  def prepare_for_display(self, user: store.User):
+    if user.is_active is None:
+      user.is_active = True
+    if user.is_admin is None:
+      user.is_admin = False
+
+  def prepare_for_save(self, user_dict: Dict, password: Optional[str]):
+    """Cleaning input user_dict."""
+    id_key = "user_id"
+    # We don't give the id to existing user to avoid SQL conflict.
+    if not user_dict.get(id_key, ""):
+      user_dict.pop(id_key, None)
+
+    user_dict["is_active"] = bool(user_dict.get("is_active", True))
+    user_dict["is_admin"] = bool(user_dict.get("is_admin", False))
```

Comment:
as far as I remember, the checkbox returns "on" if it's checked, and an empty string if not.

Reply:
not if you set its value in the html. then it becomes true/false

Reply:
but it does not help much, since it comes as a string